Collect some of scheduling metrics and scheduling throughput

ingvagabund opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

In addition to getting overall performance measurements from golang benchmark,
collect metrics that provides information about inside of the scheduling itself.

Metrics in question:
- scheduler_scheduling_algorithm_predicate_evaluation_seconds
- scheduler_scheduling_algorithm_priority_evaluation_seconds
- scheduler_binding_duration_seconds
- scheduler_e2e_scheduling_duration_seconds

Scheduling throughput is computed on the fly inside benchmarkScheduling.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #85685

**Special notes for your reviewer**:

Code for computing percentiles is copy pasted from prometheus. DataItem struct copied from clusterloader performance tests.

| name | runtime |
| ------- | ---------- |
| BenchmarkPerfScheduling | 3m4.097s |
| BenchmarkPerfSchedulingPodAntiAffinity | 5m2.016s |
| BenchmarkPerfSchedulingSecrets | 2m11.731s |
| BenchmarkPerfSchedulingInTreePVs | 32m37.221s |
| BenchmarkPerfSchedulingMigratedInTreePVs | 16m5.844s |
| BenchmarkPerfSchedulingCSIPVs | 15m9.842s |
| BenchmarkPerfSchedulingPodAffinity | 5m16.747s |
| BenchmarkPerfSchedulingNodeAffinity | 2m28.232s |

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

liu-cong commented:
/assign

ingvagabund commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -0,0 +1,356 @@
+		},
+	})
+
+	// Start measuring throughput
```
Done

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,371 @@
+
+const (
+	dateFormat                 = "2006-01-02T15:04:05Z"
+	defaultWaitForPodsInterval = 1 * time.Second
```
suggestion: `throughputSampleFrequencySeconds = 1`

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+// DataItem is the data point.
+type DataItem struct {
```
Ack.

Then can we just get rid of `summary` and use `DataItem` directly? It's more flexible too.

ingvagabund replied:
Done

liu-cong replied:
Can you resolve the comment so it's easy to track unresolved ones? Same for others. Thanks!

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,371 @@
+	}
+}
+
+func (tc *throughputCollector) collect() *summary {
```
suggestion: get rid of `summary` and return `DataItem` instead.

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,371 @@
+	}
+}
+
+func metrics2dataItems(metrics []string, labels map[string]string) []DataItem {
```
Can we create a `collector` interface and have both `prometheusCollector` and `throughputCollector` implement `collect() *DataItem` method?

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,304 @@
+package benchmark
```
Done

ingvagabund commented:
@liu-cong I moved all the collectors inside `perfScheduling` function.

ingvagabund commented:
/retest

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+	return bucketStart + (bucketEnd-bucketStart)*(rank/count)
+}
+
+// coalesceBuckets merges buckets with the same upper bound.
```
Can it happen in practice? I think the buckets come from prom Histogram metric, which shouldn't have overlapping buckets.

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+	}
+	sort.Sort(buckets)
+	buckets[len(buckets)-1].upperBound = math.Inf(+1)
+	if !math.IsInf(buckets[len(buckets)-1].upperBound, +1) {
```
I think this is unnecessary since you explicitly set the upperbound to +Inf the line above.

ingvagabund replied:
Thanks for reminding me. I moved `buckets[len(buckets)-1].upperBound = math.Inf(+1)` line before invocation of `bucketQuantile`. `bucketQuantile` is also copy-pasted from prometheus server code.

liu-cong commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+		Labels: map[string]string{
+			"Metric": pc.metric,
+		},
+		Data: map[string]float64{
```
We should have an `Average` as well.

ingvagabund replied:
What would be the average in case of a histogram with an exponential bucket length? Is `Perc50` insufficient?

liu-cong replied:
Prometheus gives you `sum` and `count`, see https://prometheus.io/docs/concepts/metric_types/#histogram.

It's useful to get the average latency of the scheduling cycle as a cross check for throughput, i.e., `average latency = 1/throughput`.

I am OK to defer this.

ingvagabund replied:
Done

ingvagabund replied:
Example:
```json
{
  "data": {
    "Average": 17.701317640933336,
    "Perc50": 13.656216216216217,
    "Perc90": 31.608961303462323,
    "Perc99": 53.44262295081967
  },
  "unit": "ms",
  "labels": {
    "Metric": "scheduler_e2e_scheduling_duration_seconds",
    "Name": "SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
  }
}
```

liu-cong commented:
Sorry for late response. I have been focusing on #86160 (the one that this PR needs to rebase on).

Once you address the comments and rebase, I will LGTM.

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+}
+
+type prometheusCollector struct {
+	metric string
```
How would you define the collector interface then? `collect() *DataItem` if used as it is needs to return a single data item. One metric for each data item. Unless we do something like:
```go
type CollectorWrapper struct {
	CollectHandler func() *DataItem
}

func (cw *CollectorWrapper) collect() *DataItem {
	return cw.CollectHandler()
}

metrics := {"metric1", ..., "metricn"}
prometheusCollector := newPrometheusCollector(metrics)
var collectors []collecter
for _, metric := range metrics {
	collectors.append(CollectorWrapper{CollectHandler: prometheusCollector.collect(metric)})
}

...
collectors.append(throughputCollector)
...

for _, coll := range collectors {
	dataItems = append(coll.collect())
}
```

liu-cong replied:
I was thinking `collect()` should return `*DataItem` instead of `*DataItem`. Will this solve the problem?

ingvagabund replied:
> should return *DataItem instead of *DataItem

Something is missing. Did you mean `collect() []*DataItem` instead?

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+			continue
+		}
+		metricFamily = mFamily
+		break
```
I prefer the former so one can first eliminate what is not relevant and process the rest.

liu-cong replied:
I don't see how the former is more efficient than the proposed. They are the same in terms of efficiency, aren't them?

ingvagabund replied:
done

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+		return nil
+	}
+
+	// clear the metrics since it's shared among all tests run inside the same binary
```
> so that next test always starts with empty prometheus metrics, right?

Yes

liu-cong replied:
can you add this to the comment?

ingvagabund replied:
done

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+func (pc *prometheusCollector) promHist2Summary(hist *dto.Histogram) *DataItem {
+	buckets := []bucket{}
+
+	if hist.SampleCount != nil && *hist.SampleCount == 0 {
```
Done

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+			}
+
+			scheduled := len(podsScheduled)
+			throughput := float64(scheduled-lastScheduledCount) / float64(float64(throughputSampleFrequency)/float64(time.Second))
```
Oh, it's a leftover, I can remove one float64 conversion: `float64(scheduled-lastScheduledCount) / (float64(throughputSampleFrequency)/float64(time.Second))`. I want to keep `float64(throughputSampleFrequency)/float64(time.Second)` explicit to provider more obvious understanding of the expression. Or:

```go
samplingRatioSeconds := float64(throughputSampleFrequency) / float64(time.Second)
throughput := float64(scheduled-lastScheduledCount) / samplingRatioSeconds
```

liu-cong commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -0,0 +1,147 @@
+			dataItems.DataItems = append(dataItems.DataItems, items...)
+		})
+	}
+	if err := dataItem2JSONFile(dataItems, b.Name()); err != nil {
```
I think the existing tests saves only one file but it doesn't have to be one. 

Can the perf dashboard parse multiple artifact files? As long as we make the perf dash happy, I don't mind one file vs. multiple files.

ingvagabund replied:
> We should just use BenchmarkPerfScheduling as the file prefix

Done. Getting `BenchmarkPerfScheduling_2020-01-08T13:40:36Z.json` like data items files now.

ingvagabund committed:
Collect some of scheduling metrics and scheduling throughput

In addition to getting overall performance measurements from golang benchmark,
collect metrics that provides information about insides of the scheduler itself.
This is a first step towards improving what we collect about the scheduler.

Metrics in question:
- scheduler_scheduling_algorithm_predicate_evaluation_seconds
- scheduler_scheduling_algorithm_priority_evaluation_seconds
- scheduler_binding_duration_seconds
- scheduler_e2e_scheduling_duration_seconds

Scheduling throughput is computed on the fly inside perfScheduling.

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -0,0 +1,348 @@
+			"Perc90":  tc.schedulingThroughputs[int(math.Ceil(float64(length*90)/100))-1],
+			"Perc99":  tc.schedulingThroughputs[int(math.Ceil(float64(length*99)/100))-1],
+		}
+		throughputSummary.Unit = "s"
```
Done

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -35,6 +37,8 @@ const (
 	throughputSampleFrequency = time.Second
 )
 
+var dataitemsdir = flag.String("dataitemsdir", "", "destination directory for storing generated data items for perf dashboard")
```
done

ingvagabund commented on test/integration/scheduler_perf/perf_util.go:
```diff
@@ -35,6 +37,8 @@ const (
 	throughputSampleFrequency = time.Second
 )
 
+var dataitemsdir = flag.String("dataitemsdir", "", "destination directory for storing generated data items for perf dashboard")
```
done

ingvagabund commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -0,0 +1,147 @@
+					items[idx].Labels = map[string]string{}
+				}
+				items[idx].Labels["Name"] = name
+			}
```
Done, moved under `perfScheduling`. The signature makes it easier to implement.

ingvagabund commented:
@liu-cong, I moved content of my original `perf_utils.go` under `utils.go` so the review comments can get a bit hard to follow and track back.

PTAL

liu-cong commented on test/integration/scheduler_perf/util.go:
```diff
@@ -66,3 +85,336 @@ func getScheduledPods(podInformer coreinformers.PodInformer) ([]*v1.Pod, error)
+	DataItems []DataItem `json:"dataItems"`
+}
+
+func dataItem2JSONFile(dataItems DataItems, namePrefix string) error {
```
nit: `dataItems2JSONFile`

ingvagabund replied:
Done

liu-cong reviewed (commented):
/lgtm

Can you share the generated `data-items` file?

ingvagabund commented:
> Can you share the generated data-items file?

Generating one with the default performance configuration. Will take a while :-|

ingvagabund commented:
I am unable to run all the tests to completion. Most likely due to https://github.com/kubernetes/kubernetes/pull/84667.

ingvagabund commented:
/retest

ingvagabund commented:
@notpad when measuring running times of the benchmark tests at [1], did you manage to run all the tests at once?

[1] https://github.com/kubernetes/kubernetes/pull/86160#issuecomment-567037587

liu-cong commented:
> I am unable to run all the tests to completion. 
I believe I ran into this too. Can you just run a subset of the tests and share an example of the JSON file generated?

ingvagabund commented:
> Can you just run a subset of the tests and share an example of the JSON file generated?

```json
{
  "version": "v1",
  "dataItems": [
    {
      "data": {
        "Average": 91.2,
        "Perc50": 104,
        "Perc90": 110,
        "Perc99": 111
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/500Nodes/500InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 1.3379603893333323,
        "Perc50": 0.7537688442211056,
        "Perc90": 3.197278911564626,
        "Perc99": 5.066666666666667
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_scheduling_algorithm_predicate_evaluation_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/500Nodes/500InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 5.029140056666668,
        "Perc50": 3.36734693877551,
        "Perc90": 12.143589743589745,
        "Perc99": 15.220512820512822
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_scheduling_algorithm_priority_evaluation_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/500Nodes/500InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 17.307066074000005,
        "Perc50": 11.805068226120857,
        "Perc90": 40.434108527131784,
        "Perc99": 59.03875968992248
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_binding_duration_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/500Nodes/500InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 25.292899956666616,
        "Perc50": 20.35409836065574,
        "Perc90": 54.48648648648649,
        "Perc99": 72.00000000000001
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_e2e_scheduling_duration_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/500Nodes/500InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 56.529411764705884,
        "Perc50": 67,
        "Perc90": 78,
        "Perc99": 82
      },
      "unit": "pods/s",
      "labels": {
        "Metric": "SchedulingThroughput",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 2.585799845066673,
        "Perc50": 1.5930031174229302,
        "Perc90": 6.418848167539267,
        "Perc99": 9.600000000000001
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_scheduling_algorithm_predicate_evaluation_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 8.593011156800031,
        "Perc50": 6.689178818112049,
        "Perc90": 15.839269406392694,
        "Perc99": 25.753222836095766
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_scheduling_algorithm_priority_evaluation_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 14.31556569933331,
        "Perc50": 9.845440494590418,
        "Perc90": 29.7982905982906,
        "Perc99": 50.021052631578954
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_binding_duration_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
      }
    },
    {
      "data": {
        "Average": 27.197705143066578,
        "Perc50": 22.882520010327912,
        "Perc90": 54.57651245551602,
        "Perc99": 72.20991253644316
      },
      "unit": "ms",
      "labels": {
        "Metric": "scheduler_e2e_scheduling_duration_seconds",
        "Name": "BenchmarkPerfScheduling/SchedulingBasic/5000Nodes/5000InitPods/1000PodsToSchedule"
      }
    }
  ]
}
```

liu-cong commented:
/lgtm

@ahg-g, can you review?

notpad commented:
> @notpad when measuring running times of the benchmark tests at [1], did you manage to run all the tests at once?
> 
> [1] [#86160 (comment)](https://github.com/kubernetes/kubernetes/pull/86160#issuecomment-567037587)

Yes, I was able to run all of the tests at once. You can try to run `ulimit -n 4096` before running the test.

ingvagabund commented on test/integration/scheduler_perf/util.go:
```diff
@@ -66,3 +85,336 @@ func getScheduledPods(podInformer coreinformers.PodInformer) ([]*v1.Pod, error)
+}
+
+func (tc *throughputCollector) run(stopCh chan struct{}) {
+	lastScheduledCount := tc.numExistingPods
```
Good point.

ingvagabund commented on test/integration/scheduler_perf/util.go:
```diff
@@ -17,15 +17,34 @@ limitations under the License.
+	"sort"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
```
data transfer object, `dto` is a common id used by https://godoc.org/github.com/prometheus/client_golang/prometheus

ingvagabund commented on test/integration/scheduler_perf/util.go:
```diff
@@ -66,3 +85,336 @@ func getScheduledPods(podInformer coreinformers.PodInformer) ([]*v1.Pod, error)
+func (b buckets) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+func (b buckets) Less(i, j int) bool { return b[i].upperBound < b[j].upperBound }
+
+// bucketQuantile calculates the quantile 'q' based on the given buckets. The
```
Yes, borrowed from https://github.com/prometheus/prometheus/blob/master/promql/quantile.go. Unfortunately, the functions are not exported.

ahg-g replied:
@liggitt are we allowed to copy code like that from another repo?

If yes, at least include a comment that this was moved from there.

liggitt replied:
I'm not sure, but I would generally expect not. @cblecker, any insight into our approach for things like this re: licensing/attribution/isolation of code like this?

cblecker replied:
Let's fork the file in a new subdir of `third_party/forked/`, and include a README file with the original filename and commit it was forked at. Then you can change these to exported functions and import/use them here.

ahg-g replied:
@ingvagabund , I think @cblecker meant that we should fork it in https://github.com/kubernetes/kubernetes/tree/master/third_party/forked

ingvagabund replied:
Done under `third_party/forked/README.md`

ahg-g replied:
@ingvagabund please see the comment above, I think you need to place the code under k/k/thrid_party/forked

ingvagabund commented on test/integration/scheduler_perf/util.go:
```diff
@@ -66,3 +85,336 @@ func getScheduledPods(podInformer coreinformers.PodInformer) ([]*v1.Pod, error)
+	for _, bckt := range hist.Bucket {
+		b := bucket{
+			count:      float64(*bckt.CumulativeCount),
+			upperBound: *bckt.UpperBound,
```
I do not. I assume the upper bound of the last bucket can be infinity, thus unset.

ingvagabund replied:
I added a check for `bckt == nil`. I don't think it will happen once a histogram is populated. Though, I can't verify it will be true in every case.

ahg-g replied:
should we coalesce the buckets here?

ingvagabund replied:
I will be very surprised if there's gonna be two buckets with the same upper bound. Dunno how it might happen given we consume the histograms directly. Unless, there's a bug in the prometheus code.

ingvagabund commented:
/test pull-kubernetes-e2e-gce

ingvagabund commented on test/integration/scheduler_perf/util.go:
```diff
@@ -66,3 +85,336 @@ func getScheduledPods(podInformer coreinformers.PodInformer) ([]*v1.Pod, error)
+		return nil
+	}
+
+	dataItem := pc.promHist2Summary(metricFamily.GetMetric()[0].GetHistogram())
```
Done

ingvagabund commented:
/test pull-kubernetes-e2e-gce

ingvagabund commented:
/test pull-kubernetes-e2e-gce-100-performance

ingvagabund commented:
/test pull-kubernetes-e2e-gce

ingvagabund commented:
/retest

ahg-g commented:
/cc @lavalamp 
Daniel, can you please review the ```third_party/forked``` changes? please see https://github.com/kubernetes/kubernetes/pull/85861#discussion_r367096687 context.

/lgtm 
/approve
for scheduler

ingvagabund commented:
/retest

lavalamp commented on third_party/forked/prometheus/prometheus/promql/prometheus_server_utils.go:
```diff
@@ -0,0 +1,162 @@
+
+func ensureMonotonic(buckets Buckets) {
+	max := buckets[0].Count
+	for i := range buckets[1:] {
```
This code neglects to check the final bucket? if len(buckets) is 10, then len(buckets[1:]) is 9, and i will range from 0 to 8. I'd expect the author wanted it to go from 1 to 9.

lavalamp replied:
(the final bucket is ignored in the search so it looks like this doesn't matter?)

lavalamp commented on third_party/forked/prometheus/prometheus/promql/prometheus_server_utils.go:
```diff
@@ -0,0 +1,162 @@
+//
+// This is because scraped data is not made available to rule evaluation or
+// federation atomically, so some buckets are computed with data from the
+// most recent scrapes, but the other buckets are missing data from the most
```
Reading this code definitely reduced my faith in prometheus metrics. :/

lavalamp commented:
@kubernetes/dep-approvers for the thirdparty forked code. I can approve if they can't for some reason.

We have done things like this in the past when there wasn't another option.

Are you sure that code is so great that you want to reuse it? :upside_down_face: It looks like it papers over data collection errors to me.

lavalamp commented:
It looks like we stopped using the `forked/` subdir, and that we haven't substantially changed anything in it for 2 years, so let's definitely have @kubernetes/dep-approvers comment on this approach. They should probably put themselves in OWNERS here.

BenTheElder commented:
@lavalamp it's not possible for @kubernetes/dep-approvers to approve fully :(
go.mod / go.sum are in the repo root

BenTheElder commented:
Can we please put a note in the PR body about why we're forking something?

ingvagabund commented:
> Can we please put a note in the PR body about why we're forking something?

Do you mean (https://github.com/kubernetes/kubernetes/pull/85861/commits/c5e97fff18b028c819967d8f9ebe60c4bcc99fa7):
```
Fork and export code from github.com/prometheus/prometheus/promql

It's not good practice to just copy-paste a code from a third party repository.
Instead, it's forked, needed parts are exported and the corresponding repository
location and code revision written down.
```

Or something more specific?

BenTheElder commented:
forked is standard practice vs copy paste, I'm missing why we're forking or copying it to begin with?

@lavalamp third_party is definitely still in use, though we also have some libraries for reuse that are forked into their own repos like klog. those repos / third_party are pretty much only used for things we couldn't reasonably get done upstream, which is relatively rare (a decent amount of it is golang internals...)

ingvagabund commented:
> forked is standard practice vs copy paste, I'm missing why we're forking or copying it to begin with?

I did not want to re-implement what's already provided by prometheus. Also, I wanted exactly the same implementation for consistency so we get the same number no matter if we generate the metrics directly (as I do) or use prometheus query.

BenTheElder commented:
>I did not want to re-implement what's already provided by prometheus. Also, I wanted exactly the same implementation for consistency so we get the same number no matter if we generate the metrics directly (as I do) or use prometheus query.

why can't we vendor it?

ahg-g commented:
> > I did not want to re-implement what's already provided by prometheus. Also, I wanted exactly the same implementation for consistency so we get the same number no matter if we generate the metrics directly (as I do) or use prometheus query.
> 
> why can't we vendor it?

Because those functions are not exported.

BenTheElder commented:
in the past we've requested functionality be made more accessible before
forking

On Mon, Jan 27, 2020 at 10:15 AM ahg-g <notifications@github.com> wrote:

> I did not want to re-implement what's already provided by prometheus.
> Also, I wanted exactly the same implementation for consistency so we get
> the same number no matter if we generate the metrics directly (as I do) or
> use prometheus query.
>
> why can't we vendor it?
>
> Because those functions are not exported.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/85861?email_source=notifications&email_token=AAHADKZB3A5UKXZOMSI6LWDQ74QDRA5CNFSM4JU6JYY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKAQKIY#issuecomment-578880803>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHADK5VCGZXGPM7MPPIWN3Q74QDRANCNFSM4JU6JYYQ>
> .
>

ingvagabund commented:
/retest

haosdent commented:
/retest

ingvagabund commented:
/retest

ingvagabund commented:
@BenTheElder @lavalamp I re-implemented the percentile computation and dropped forked prometheus completely.

ahg-g reviewed (commented):
/lgtm

Thanks!

ingvagabund commented:
/retest

ingvagabund commented:
/retest

ingvagabund commented:
/retest

ingvagabund commented:
/retest

ingvagabund commented:
/retest

ingvagabund commented:
/test pull-kubernetes-e2e-gce

ingvagabund commented:
/retest

ingvagabund commented:
/test pull-kubernetes-e2e-gce-100-performance

mikedanese commented on test/integration/scheduler_perf/BUILD:
```diff
@@ -20,7 +20,10 @@ go_library(
         "//staging/src/k8s.io/client-go/rest:go_default_library",
+        "//staging/src/k8s.io/component-base/metrics/legacyregistry:go_default_library",
         "//test/integration/util:go_default_library",
+        "//vendor/github.com/prometheus/client_model/go:go_default_library",
```
This dependency is breaking the build. See https://github.com/kubernetes/kubernetes/pull/84302.

Unclear why this wasn't caught by the presubmit... cc @BenTheElder

mikedanese replied:
I opened https://github.com/kubernetes/test-infra/pull/16168 to add coverage. Apparently we miss this in the build presubmit.

liggitt replied:
can we revert this to unbreak master, then reintroduce when fixed?

ingvagabund replied:
@serathius so what's the right way to import `github.com/prometheus/client_model/go` ? I am going through #84302 and what's link in there though it may take some time to figure out what to do.

serathius replied:
You should not use or depend on internals of prometheus. 
Please move functions that rely on prometheus internals into staging/src/k8s.io/component-base/metrics/testutil
Example PR https://github.com/kubernetes/kubernetes/pull/85289

ingvagabund replied:
@serathius @liggitt @mikedanese Moving the code under `staging/src/k8s.io/component-base/metrics/testutil` in https://github.com/kubernetes/kubernetes/pull/87923
